src/execution_plan/optimizations/traverse_order.c

```diff
@@ -5,74 +5,229 @@
+                if(!exp->operands[k].transpose) penalty += T;
+            }
         }
     }
 
-    // See if there's a filter applied to the last expression.
-    exp = exps[expCount-1];
-    lastExpLabeled = exp->src_node->label || exp->dest_node->label;
-    destAlias = exp->dest_node->alias;
-    srcAlias = exp->src_node->alias;
-
-    for(int i = 0; i < aliasesCount; i++) {
-        char *alias;
-        Vector_Get(aliases, i, &alias);
-        if(strcmp(alias, srcAlias) == 0 || strcmp(alias, destAlias) == 0) {
-            order = TRAVERSE_ORDER_LAST;
-            goto cleanup;
+    return penalty;
+}
+
+static int reward_arrangement(Arrangement arrangement, uint exp_count, const FT_FilterNode *filters) {
+    // Arrangement_Print(arrangement, exp_count);
+    int reward = 0;
+    rax *filtered_entities = FilterTree_CollectAliases(filters);
+
+    // A bit naive at the moment.
+    for(uint i = 0; i < exp_count; i++) {
+        AlgebraicExpression *exp = arrangement[i];
+        char *src_alias = exp->src_node->alias;
+        char *dest_alias = exp->dest_node->alias;
+
+        if(raxFind(filtered_entities, (unsigned char*)src_alias, strlen(src_alias)) != raxNotFound) {
+            reward += F * (exp_count - i);
+            raxRemove(filtered_entities, (unsigned char*)src_alias, strlen(src_alias), NULL);
+        }
+        if(raxFind(filtered_entities, (unsigned char*)dest_alias, strlen(dest_alias)) != raxNotFound) {
+            reward += F * (exp_count - i);
+            raxRemove(filtered_entities, (unsigned char*)dest_alias, strlen(dest_alias), NULL);
         }
     }
 
-    /* If we're here, there are no filters on either the first or the last
-     * expression. The next-best criteria for choosing traversal order is
-     * to prefer an expression in which the source or destination has a label,
-     * as label scans are significantly faster than scanning all nodes.
-     * If both expressions use labels (or neither do), we'll default
-     * to TRAVERSE_ORDER_FIRST. */
-    if(!firstExpLabeled && lastExpLabeled) order = TRAVERSE_ORDER_LAST;
-
-cleanup:
-    for(int i = 0; i < aliasesCount; i++) {
-        char *alias;
-        Vector_Get(aliases, i, &alias);
-        free(alias);
+    // printf("reward: %d\n", reward);
+    raxFree(filtered_entities);
+    return reward;
+}
+
+static int score_arrangement(Arrangement arrangement, uint exp_count, const FT_FilterNode *filters) {
+    int score = 0;
+    int penalty = penalty_arrangement(arrangement, exp_count);
+    int reward = reward_arrangement(arrangement, exp_count, filters);
+    score -= penalty;
+    score += reward;
+    return score;
+}
+
+/* Given a set of algebraic expressions representing a graph traversal 
+ * we pick the order in which the expressions will be evaluated 
+ * taking into account filters and transposes. 
+ * exps will reordered. */
+void orderExpressions(AlgebraicExpression **exps, uint exps_count, const FT_FilterNode *filters) {
```

Comment:
We've already talked about the logic here and later optimizations we could make, so I'm not going to comment over this entire file. What's written here looks fine to me!

Reply:
👍